layers/gre.go

```diff
@@ -19,12 +19,15 @@ type GRE struct {
 	Protocol                                                                   EthernetType
 	Checksum, Offset                                                           uint16
 	Key, Seq                                                                   uint32
-	*GRERouting
+	GRERouting                                                                 GRERouting
 }
 
 // GRERouting is GRE routing information, present if the RoutingPresent flag is
 // set.
-type GRERouting struct {
+type GRERouting []*GRESourceRouteEntry
```

Comment:
This type switching here is backwards-incompatible to anyone who's already using this code.  How about instead of a slice, you make GRERouting a linked list?  Basically, just add a

    Next *GRERouting

field to the existing struct, which is nil if it's the last one and non-nil if there's more.